katana/native/cases/static/cases/js/casesApp.js

```diff
@@ -0,0 +1,1040 @@
+},
+
+
+// Get Results directory from the case on a popup dialog
+// The dialog is shown BELOW the popup. BUG. 
+	getResultsDirForCaseStep: function() {
+	  var tag = "#StepInputDataFile";
+      var callback_on_accept = function(selectedValue) { 
+      		console.log(selectedValue);
+      		var popup = katana.$activeTab.find("#editCaseStepDiv").attr('popup-id');
+
+      		// Convert to relative path.
+      		var pathToBase = katana.$activeTab.find('#savefilepath').text();
+      		console.log("File path ==", pathToBase);
+      		// var nf = katana.fileExplorerAPI.prefixFromAbs(pathToBase, selectedValue);
+      		var nf = prefixFromAbs(pathToBase, selectedValue);
+      		
+      		popup.find(tag).attr("value", nf);
+      		popup.find(tag).attr("fullpath", selectedValue);
+
+            };
+      var callback_on_dismiss =  function(){ 
+      		console.log("Dismissed");
+	 };
+	 //var popup = katana.$activeTab.find("#editCaseStepDiv").attr('popup-id');
+     katana.fileExplorerAPI.openFileExplorer("Select a file", false , $("[name='csrfmiddlewaretoken']").val(), false, callback_on_accept, callback_on_dismiss);
+},
+
+				
+// Fills in the drop down based on the number of steps you have 
+// in the case. 
+
+	 fillCaseDefaultGoto: function() {
+	var gotoStep = katana.$activeTab.find('#default_onError').val();
+	console.log("Step ", gotoStep);
+	var defgoto = katana.$activeTab.find('#default_onError_goto'); 
+		defgoto.hide();
+
+	if (gotoStep.trim() == 'goto'.trim()) { 
+		defgoto.show();
+	} else {
+		defgoto.hide();
+		
+	}
+
+	defgoto.empty(); 
+	var xdata = caseApp.jsonCaseObject["Steps"]['step']; // ['Testcase']; 
+	console.log("Step....",Object.keys(xdata).length, xdata);
+	for (var s=0; s<Object.keys(xdata).length; s++ ) {
+		defgoto.append($('<option>',{ value: s,  text: s+1}));
+	}
+},
+
+// Saves the UI to memory and sends to server as a POST request
+	sendCaseToServer: function () {
+	caseApp.mapUiToCaseJson();
+	var url = "./cases/getCaseDataBack";
+	var csrftoken = $("[name='csrfmiddlewaretoken']").attr('value');
+
+	$.ajaxSetup({
+			function(xhr, settings) {
+            xhr.setRequestHeader("X-CSRFToken", csrftoken)
+    	}
+	});
+
+	var topNode  = { 'Testcase' : caseApp.jsonCaseObject};
+
+	$.ajax({
+    url : url,
+    type: "POST",
+    data : { 
+    	'json': JSON.stringify(topNode),	
+    	'filetosave': katana.$activeTab.find('#my_file_to_save').attr('value'),
+    	'savesubdir': katana.$activeTab.find('#savesubdir').text(),
+    	},
+    headers: {'X-CSRFToken':csrftoken},
+    //contentType: 'application/json',
+    success: function( data ){
+        alert("Sent");
+    	}
+	});
+},
+
+// Case steps that are incomplete in an XML file can cause havoc on the UI. 
+// Fill in defaults for missing fields. 
+	fillStepDefaults: function(oneCaseStep) {
+
+		if (! oneCaseStep['step']){
```

Comment:
The check and add for this js function should move to server side
Can use a template json on the server side
Once the server side validation is good enough, this part can be removed

Reply:
I prefer checks and balances on the client even if we have checks on the server because the errors in javascript are catastrophic